src/workerd/api/pyodide/pyodide.c++

```diff
@@ -102,6 +116,176 @@ int ArtifactBundler::readMemorySnapshot(int offset, kj::Array<kj::byte> buf) {
   return readToTarget(KJ_REQUIRE_NONNULL(existingSnapshot), offset, buf);
 }
 
+kj::Array<kj::String> ArtifactBundler::parsePythonScriptImports(kj::Array<kj::String> files) {
+  auto result = kj::Vector<kj::String>();
+
+  for (auto& file: files) {
+    // Returns the number of characters skipped. When `oneOf` is not found, skips to the end of
+    // the string.
+    auto skipUntil = [](kj::StringPtr str, std::initializer_list<char> oneOf, int start) -> int {
+      int result = 0;
+      while (start + result < str.size()) {
+        char c = str[start + result];
+        for (char expected: oneOf) {
+          if (c == expected) {
+            return result;
+          }
+        }
+
+        result++;
+      }
+
+      return result;
+    };
+
+    // Skips while current character is in `oneOf`. Returns the number of characters skipped.
+    auto skipWhile = [](kj::StringPtr str, std::initializer_list<char> oneOf, int start) -> int {
+      int result = 0;
+      while (start + result < str.size()) {
+        char c = str[start + result];
+        bool found = false;
+        for (char expected: oneOf) {
+          if (c == expected) {
+            result++;
+            found = true;
+            break;
+          }
+        }
+
+        if (!found) {
+          break;
+        }
+      }
+
+      return result;
+    };
+
+    // Skips one of the characters (specified in `oneOf`) at the current position. Otherwise
+    // throws. Returns the number of characters skipped.
+    auto skipChar = [](kj::StringPtr str, std::initializer_list<char> oneOf, int start) -> int {
+      for (char expected: oneOf) {
+        if (str[start] == expected) {
+          return 1;
+        }
+      }
+
+      KJ_FAIL_REQUIRE("Expected ", oneOf, "but received", str[start]);
+    };
+
+    auto parseKeyword = [](kj::StringPtr str, kj::StringPtr ident, int start) -> bool {
+      int i = 0;
+      for (; i < ident.size() && start + i < str.size(); i++) {
+        if (str[start + i] != ident[i]) {
+          return false;
+        }
+      }
+
+      return i == ident.size();
+    };
+
+    // Returns the size of the import identifier or 0 if no identifier exists at `start`.
+    auto parseIdent = [](kj::StringPtr str, int start) -> int {
+      // https://docs.python.org/3/reference/lexical_analysis.html#identifiers
+      //
+      // We also accept `.` because import idents can contain it.
+      if (std::isdigit(str[start])) {
+        return 0;
+      }
+      int i = 0;
+      for (; start + i < str.size(); i++) {
+        char c = str[start + i];
+        bool validIdentChar = std::isalpha(c) || std::isdigit(c) || c == '_' || c == '.';
```

Comment:
This would presumably accept `foo..bar` but consecutive periods shouldn't be allowed?

Reply:
yeah, but again shouldn't be possible in valid Python

Reply:
Yeah, I just didn't know this was an assumption that could be made. I think this should definitely be called out in the method's documentation.